lib/auth/in_memory/checker.js

```diff
@@ -0,0 +1,176 @@
+ */
+function checkExists(data, container, name, type, obj) {
+    if (obj[name] === undefined) {
+        data.errors.push({
+            txt: 'missing property in auth entity',
+            obj: {
+                entity: container,
+                property: name,
+            },
+        });
+        return false;
+    }
+    return checkType(data, container, name, type, obj);
+}
+
+function checkUser(data, userObj) {
+    if (checkExists(data, 'User', 'arn', 'string', userObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.arns[userObj.arn] = incr(data.arns[userObj.arn]);
+    }
+    if (checkExists(data, 'User', 'email', 'string', userObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.emails[userObj.email] = incr(data.emails[userObj.email]);
+    }
+    if (checkExists(data, 'User', 'keys', 'array', userObj)) {
+        userObj.keys.forEach(keyObj => {
+            // eslint-disable-next-line no-param-reassign
+            data.keys[keyObj.access] = incr(data.keys[keyObj.access]);
+        });
+    }
+}
+
+function checkAccount(data, accountObj) {
+    if (checkExists(data, 'Account', 'email', 'string', accountObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.emails[accountObj.email] = incr(data.emails[accountObj.email]);
+    }
+    if (checkExists(data, 'Account', 'arn', 'string', accountObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.arns[accountObj.arn] = incr(data.arns[accountObj.arn]);
+    }
+    if (checkExists(data, 'Account', 'canonicalID', 'string', accountObj)) {
+        // eslint-disable-next-line no-param-reassign
+        data.canonicalIds[accountObj.canonicalID] =
+            incr(data.canonicalIds[accountObj.canonicalID]);
+    }
+
+    if (accountObj.users) {
+        if (checkType(data, 'Account', 'users', 'array', accountObj)) {
+            accountObj.users.forEach(userObj => checkUser(data, userObj));
+        }
+    }
+
+    if (accountObj.keys) {
+        if (checkType(data, 'Account', 'keys', 'array', accountObj)) {
+            accountObj.keys.forEach(keyObj => {
+                // eslint-disable-next-line no-param-reassign
+                data.keys[keyObj.access] = incr(data.keys[keyObj.access]);
+            });
+        }
+    }
+}
+
+function dumpCountError(property, obj) {
+    let count = 0;
+    Object.keys(obj).forEach(key => {
+        if (obj[key] > 1) {
+            log.error('property should be unique', {
+                property,
+                value: key,
+                count: obj[key],
+            });
+            ++count;
+        }
+    });
+    return count;
+}
+
+function dumpErrors(checkData) {
+    let nerr = dumpCountError('CanonicalID', checkData.canonicalIds);
+    nerr += dumpCountError('Email', checkData.emails);
+    nerr += dumpCountError('ARN', checkData.arns);
+    nerr += dumpCountError('AccessKey', checkData.keys);
+
+    if (nerr > 0) {
+        return true;
+    }
+
```

Comment:
If I'm reading this correctly, we log errors for duplicate items and then don't log any other formatting errors if there are duplicates?  Why?  Also, a few more comments would be helpful here.